### ISSUE TITLE
Runtime Issues caused by Nesting

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#1116]
In preparation of validating the improved *Core Type System* https://github.com/dagger/dagger/pull/1117, here's a recollection of problems we've had with the current runtime due to *nesting*.

## What is nesting?

Nesting is the practice of passing data from a pipeline to another by referencing the pipeline itself.

For instance, an `#Artifact` (a compute pipeline, e.g. something with `#up`) can mount another `#Artifact` by passing a reference to it:

```
a: #up: [
  op.#FetchContainer & { ref: "alpine" }
]

b: #up: [
  op.#Exec & { mount: "/a": from: a }
]
```

Note the nesting in `from: a`. This will `cue eval` (e.g. what the runtime sees) to:

```
a: {
    #up: [{
        do:  "fetch-container"
        ref: "alpine"
    }]
}
b: {
    #up: [{
        do: "exec"
        mount: {
            "/a": {
                from: {
                    #up: [{
                        do:  "fetch-container"
                        ref: "alpine"
                    }]
                }
            }
        }
    }]
}
```

From the runtime's perspective, `b.#up[0].mount["/a"].from` and `a` are two distinct tasks and are indistinguishable from each other (see *Annex: Why can't we follow references with nesting?*)

## Problems

* **Logs and Error Messages**
  * Because of how the CUE tree looks like when using nesting, we routinely see errors and log messages with CUE paths such as `a.foo.#up[0].from.deploy.#up[3].mount["/src"]`
  * See one example in #613 
    * `site.netlify.contents.build.from.command: 2 errors in empty disjunction`
    * Here `site.netlify.contents` is an `#Artifact` reference, pointing to a value somewhere
    * Within this reference, there's another nested reference in `build.from` pointing yet to a different value
    * And finally, after 2 levels of indirection, there's a disjunction problem in the `command` field
    * Without nesting, we'd see the actual CUE path of where the problem is (e.g. `website.command: empty disjunction`)
* **Flaky CUE paths in logs and errors**
  * Let's say we have a `build: yarn.#Package` in the plan. Logs and errors will show up as `build`
  * Now we add a `deploy: netlify.#Site & { contents: build }`
  * When running `dagger up` again, now `build` **may** show up either as `deploy.contents` or `build`
  * This is because for cueflow, both `build` and `deploy.contents` are two different *tasks*
  * `deploy.contents` is a dependency of `deploy` whereas `build` is not. Both will be executed in parallel.
  * The CUE path in the logs/errors will be whichever got to execute first (undefined)
* **No Reliable CUE Path**
  * With nested tasks, we cannot rely on a *task*'s CUE path in the code base as a stable identifier
  * That's because reference checking doesn't work with nesting (see *Annex*)
  *  Injecting temporary **IDs** in the CUE tree mitigates the problem by providing a stable identifier for the duration of `dagger up`
  * That's enough for caching, but will it be enough for future use cases? For instance state management (a.k.a. `dagger down`)
* **CUE Performance**
  * Nested definitions (e.g. `mount` takes an `#Artifact` which itself provides `mount` which itself provides an `#Artifact`, ...) have caused various performance issues (i.e. *billion laughs*).
  * This led to all `#Artifact` references being replaced by `_` as a workaround
  * Even then, we routinely suffer from slow evaluation times (e.g. `cue eval`) and slow fill times (e.g. `cue.Fill()`)
    * References needed here
    * IIRC @grouville recently struggled with a large config taking minutes of evaluation between each task
* **Confused CUE Type System**
  * The *open definition* (e.g. `{ ... _ }`) of `#Artifact` has caused problems
  * Since anything is technically an artifact, it confuses CUE's type system (e.g. `if (source & string) != _|_`)
  * Passing a wrong argument (or forgetting to provide it) causes hard to understand CUE errors
  * References needed here -- @samalba I remember you struggled with this
* **Caching Issues** (#914)
  * References are seen as a different *task* by *cueflow*
  * We rely on BuildKit for deduplication, avoiding unnecessary reference computation. However it is limited to LLB instructions.
    * non-LLB operations (e.g. gRPC calls to read files, call to external frontends, pushing images, etc) are **not** de-duplicated (see #449 #591)
    * non-idempotent operations will compile to a different LLB state each time they are referenced, defeating BuildKit cache. For instance, operations using the current *cue path* (e.g. cache mounts #495) and operations using randomness (e.g. `always: true` which is re-computed at each reference #42)
* **Misc**
  * The output of `cue eval` / `dagger query` is hard to understand by a human because of the nesting of values
  * Listing *tasks* ahead of time (e.g. future `dagger plan`?) will show up duplicates
  * Debugging the runtime can be painful at times

## Annex: Why can't we follow references with nesting?

Consider this simple nested task definition, which takes an `exec` command to run and can `mount` another task:

```
#Task: {
    mount: #Task
    exec: string
}

a: #Task & {
    mount: #Task & {
        mount: #Task & {
            exec: "foo"
        }
    }
}

b: a
```

With cueflow, we'll "find" both `a.mount.mount.exec` and `b.mount.mount.exec`, however figuring out the canonical path (`a.mount.mount.exec`) is tricky:

```
b.mount.mount.exec --> reference: false
b.mount.mount --> reference: false
b.mount --> reference: false
b --> reference to a
```

Going from `b.mount.mount.exec` to `a.mount.mount.exec` is challenging.

This is the problem in its simplest form. When you add expressions (interpolation, disjunctions, ...) and multiple layers of indirection (a to b to c) it becomes extremely hard to follow references.

It's been attempted multiple times without success (see https://github.com/dagger/dagger/pull/922)
